ArjunaJTS/jtax/classes/com/arjuna/ats/internal/jta/transaction/jts/jca/XATerminatorImple.java

```diff
@@ -506,9 +441,107 @@ public void removeImportedTransaction(Xid xid) throws XAException {
         return xidsToRecover.toArray(NO_XIDS);
     }
 
+    /**
+     * Recovering /JCA section of object store filtering on xid or parentNodeName.
+     *
+     * This JTS variant of this method does not accept parentNodeName as can't be filtered over it,
+     * as JTS does not encode this information into {@link Xid}.
+     *
+     * Filtering over Xid is run without knowledge of parentNodeName.
+     *
+     * Expected to be used with null parameters provided when all /JCA subordinate records
+     * which are part of the object store are returned.
+     * 
+     * @param xid
+     * @param parentNodeName  has to be null
+     * @return array of subordinate recovered xids
+     */
     @Override
-    public Xid[] doRecover(Xid xid, String parentNodeName) throws XAException, NotSupportedException {
-        throw new NotSupportedException();
+    public Xid[] doRecover(Xid xid, String parentNodeName) throws XAException {
+        Xid[] indoubt = null;
+
+        try
+        {
+            RecoveryStore recoveryStore = StoreManager.getRecoveryStore();
+            InputObjectState states = new InputObjectState();
+
+            // only look in the JCA section of the object store
+            if (recoveryStore.allObjUids(ServerTransaction.getType(), states) && (states.notempty()))
+            {
+                Stack<Transaction> values = new Stack<Transaction>();
+                boolean finished = false;
+
+                do
+                {
+                    Uid uid = null;
+
+                    try
+                    {
+                        uid = UidHelper.unpackFrom(states);
+                    }
+                    catch (IOException ex)
+                    {
+                        jtsLogger.i18NLogger.info_fail_to_read_subordinate_uid(recoveryStore, states, ex);
+
+                        finished = true;
+                    }
+
+                    if (uid.notEquals(Uid.nullUid()))
+                    {
+                        if (parentNodeName != null)
+                        {
+                            // JTS uid does not encode parent node name thus we can't filter on it, see JBTM-2124
+                            throw new IllegalArgumentException("JTS doRecover method does not handle parentNodeName, see JBTM-2124");
+                        }
+                        else if (xid == null)
+                        {
+                            Transaction tx = SubordinationManager.getTransactionImporter().recoverTransaction(uid);
+
+                            if (tx != null)
+                                values.push(tx);
+                        }
+                        else // parentNodeName == null && tx != null
+                        {
+                            SubordinateAtomicTransaction sat = new SubordinateAtomicTransaction(uid);
```

Comment:
I was think it is fine to throw the exception because iiuc it would never be called except with nulls?